### PR TITLE
Extend `restart` to handle a restaged app

### LINKED ISSUE #2220: Restart/Restage do not effectively apply env vars
## Issue.
Once an app is deployed, applying the env vars later with restart/restage seem not to have an efect on making the app work.

So this will deploy a fully working app:
```
epinio push --name testapp-gituburl-0 --git https://github.com/epinio/example-go,main --env='BP_KEEP_FILES=static/*'
```

However if deploy without the en vars first and later add them it wont:

```
epinio push --name testapp-gituburl --git https://github.com/epinio/example-go,main
epinio app env set testapp-gituburl BP_KEEP_FILES 'static/*'
epinio app restage testapp-gituburl 
```

The details show the env var is there but it returns a 404 on the UI:
```
🚢  Show application details
Namespace: workspace
Application: testapp-gituburl

✔️  Details:
|         KEY          |                                            VALUE                                            |
|----------------------|---------------------------------------------------------------------------------------------|
| Origin               | https://github.com/epinio/example-go                                                        |
|                      | @ main                                                                                      |
| Created              | 2023-04-18 17:28:02 +0200 CEST                                                              |
| Status               | 1/1                                                                                         |
| Username             | admin                                                                                       |
| Running StageId      | bbe8c56067eeb990                                                                            |
| Last StageId         | 1706ef1b1da34745                                                                            |
| Age                  | 1m39s                                                                                       |
| Internal Route       | rtestapp-gituburl-3609caf5e8714c257de51705edff8e44d5a20d36.workspace.svc.cluster.local:8080 |
| Active Routes        |                                                                                             |
|                      | testapp-gituburl.192.168.16.3.omg.howdoi.website                                            |
| App Chart            | standard                                                                                    |
| Desired Instances    |                                                                                           1 |
| Bound Configurations |                                                                                             |
| Environment          |                                                                                             |
|   - BP_KEEP_FILES    | static/*                                                                                    |
| Chart Values         |                                                                                             |

✔️  Instances: 
|                              NAME                               | READY | MEMORY  | MILLICPUS | RESTARTS |  AGE  |
|-----------------------------------------------------------------|-------|---------|-----------|----------|-------|
| rtestapp-gituburl-3609caf5e8714c257de51705edff8e44d5a20d36xffm5 | true  | 1.8 MiB |         0 |        0 | 1m39s |

```

Image

### PR DESCRIPTION
**Pull request by andreas-kupries**
Fix #2220

### CODE REVIEW

File: internal/api/v1/application/restart.go [hunk 1]
```diff
@@ -49,6 +51,31 @@ func Restart(c *gin.Context) apierror.APIErrors {
...
+		pieces := strings.Split(app.ImageURL, ":")
+		pieces[len(pieces)-1] = app.StageID
+		newImageURL := strings.Join(pieces, ":")
+
+		// .. and save it for `DeployApp` to find.
+
+		appRef := models.NewAppRef(appName, namespace)
+		applicationCR, err := application.Get(ctx, cluster, appRef)
+		if err != nil {
+			return apierror.InternalError(err, "failed to get the application resource")
```
Comment (enrichman) [inline internal/api/v1/application/restart.go:71]:
```suggestion
			return apierror.InternalError(err, "getting the application resource")
```

File: internal/api/v1/application/restart.go [hunk 2]
```diff
@@ -49,6 +51,31 @@ func Restart(c *gin.Context) apierror.APIErrors {
...
+		}
+		err = deploy.UpdateImageURL(ctx, cluster, applicationCR, newImageURL)
+		if err != nil {
+			return apierror.InternalError(err, "failed to set application's image url")
```
Comment (enrichman) [inline internal/api/v1/application/restart.go:75]:
```suggestion
			return apierror.InternalError(err, "updating application's image url")
```

**Review by enrichman — APPROVED** [review]
LGTM, but I left a couple of suggestions that I found interesting while reading this article:

https://preslav.me/2023/04/14/golang-error-handling-is-a-form-of-storytelling/#crafting-a-good-message-is-key

**Comment by mmartin24** [comment]
Tested the pr and worked. Just mentioned a question here: https://github.com/epinio/epinio/issues/2220#issuecomment-1516136638

**Review by enrichman — APPROVED** [review]

**Pull request merged**